File: scripts/gen-geo.mjs
/**
 * 從統一資訊提供的「縣市區域範例資料.json」產生 TS 常數檔。
 * 該檔案是多個 JSON 物件串接（{"sys_county":[...]} 後面接 {"sys_district":[...]}），
 * 不是合法的單一 JSON，所以這裡用括號配對切開後逐段 parse。
 *
 * 用法：node scripts/gen-geo.mjs
 */
import { readFileSync, writeFileSync, readdirSync } from 'node:fs';
import { join } from 'node:path';

const dataDir = readdirSync('.').find((d) => d.includes('命題數據集'));
if (!dataDir) throw new Error('找不到命題數據集資料夾');
const src = join(dataDir, '縣市區域範例資料.json');
const raw = readFileSync(src, 'utf8');

/** 用大括號配對，把串接的多個 JSON 物件切出來 */
function splitJsonObjects(text) {
  const out = [];
  let depth = 0;
  let start = -1;
  let inStr = false;
  let esc = false;
  for (let i = 0; i < text.length; i++) {
    const c = text[i];
    if (inStr) {
      if (esc) esc = false;
      else if (c === '\\') esc = true;
      else if (c === '"') inStr = false;
      continue;
    }
    if (c === '"') inStr = true;
    else if (c === '{') {
      if (depth === 0) start = i;
      depth++;
    } else if (c === '}') {
      depth--;
      if (depth === 0 && start >= 0) {
        out.push(text.slice(start, i + 1));
        start = -1;
      }
    }
  }
  return out;
}

const counties = [];
const districts = [];
for (const chunk of splitJsonObjects(raw)) {
  let obj;
  try {
    obj = JSON.parse(chunk);
  } catch {
    continue;
  }
  if (Array.isArray(obj.sys_county)) counties.push(...obj.sys_county);
  if (Array.isArray(obj.sys_district)) districts.push(...obj.sys_district);
}

const activeCounties = counties
  .filter((c) => c.is_deleted === '0')
  .sort((a, b) => a.sort - b.sort)
  .map((c) => ({ code: c.code, name: c.name }));

const activeDistricts = districts
  .filter((d) => d.is_deleted === '0')
  .sort((a, b) => (a.county_code === b.county_code ? a.sort - b.sort : a.county_code.localeCompare(b.county_code)))
  .map((d) => ({ code: d.code, countyCode: d.county_code, name: d.name, zip: d.zip }));

const banner = `/**
 * 自動產生，請勿手改。來源：${src.replace(/\\/g, '/')}
 * 重新產生：node scripts/gen-geo.mjs
 * sys_county / sys_district（縣市代碼 2 碼、行政區代碼 3 碼）
 */`;

const ts = `${banner}

export interface CountyRow {
  code: string;
  name: string;
}

export interface DistrictRow {
  code: string;
  countyCode: string;
  name: string;
  zip: string;
}

export const COUNTIES: CountyRow[] = ${JSON.stringify(activeCounties, null, 2)};

export const DISTRICTS: DistrictRow[] = ${JSON.stringify(activeDistricts, null, 2)};
`;

const outPath = join('packages', 'backend', 'src', 'data', 'geo.generated.ts');
writeFileSync(outPath, ts, 'utf8');
console.log(`counties=${activeCounties.length} districts=${activeDistricts.length} -> ${outPath}`);

File: scripts/gen_geo.py
"""從統一資訊提供的「縣市區域範例資料.json」產生 Python 常數檔。

該檔案是多個 JSON 物件串接（{"sys_county":[...]} 後面接 {"sys_district":[...]}），
不是合法的單一 JSON，所以用大括號配對切開後逐段 parse。

用法：python scripts/gen_geo.py
"""

from __future__ import annotations

import json
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
OUT = ROOT / "packages" / "api" / "op_agent" / "geo_generated.py"


def find_dataset_dir() -> Path:
    for p in ROOT.iterdir():
        if p.is_dir() and "命題數據集" in p.name:
            return p
    raise SystemExit("找不到命題數據集資料夾")


def split_json_objects(text: str) -> list[str]:
    """用大括號配對，把串接的多個 JSON 物件切出來。"""
    out: list[str] = []
    depth = 0
    start = -1
    in_str = False
    esc = False
    for i, c in enumerate(text):
        if in_str:
            if esc:
                esc = False
            elif c == "\\":
                esc = True
            elif c == '"':
                in_str = False
            continue
        if c == '"':
            in_str = True
        elif c == "{":
            if depth == 0:
                start = i
            depth += 1
        elif c == "}":
            depth -= 1
            if depth == 0 and start >= 0:
                out.append(text[start : i + 1])
                start = -1
    return out


def main() -> None:
    src = find_dataset_dir() / "縣市區域範例資料.json"
    raw = src.read_text(encoding="utf-8")

    counties: list[dict] = []
    districts: list[dict] = []
    for chunk in split_json_objects(raw):
        try:
            obj = json.loads(chunk)
        except json.JSONDecodeError:
            continue
        if isinstance(obj.get("sys_county"), list):
            counties.extend(obj["sys_county"])
        if isinstance(obj.get("sys_district"), list):
            districts.extend(obj["sys_district"])

    active_counties = [
        {"code": c["code"], "name": c["name"]}
        for c in sorted(
            (c for c in counties if c.get("is_deleted") == "0"),
            key=lambda c: c["sort"],
        )
    ]
    active_districts = [
        {
            "code": d["code"],
            "county_code": d["county_code"],
            "name": d["name"],
            "zip": d["zip"],
        }
        for d in sorted(
            (d for d in districts if d.get("is_deleted") == "0"),
            key=lambda d: (d["county_code"], d["sort"]),
        )
    ]

    body = f'''"""自動產生，請勿手改。

來源：{src.relative_to(ROOT).as_posix()}
重新產生：python scripts/gen_geo.py
對應 sys_county / sys_district（縣市代碼 2 碼、行政區代碼 3 碼）
"""

from __future__ import annotations

from typing import TypedDict


class CountyRow(TypedDict):
    code: str
    name: str


class DistrictRow(TypedDict):
    code: str
    county_code: str
    name: str
    zip: str


COUNTIES: list[CountyRow] = {json.dumps(active_counties, ensure_ascii=False, indent=4)}

DISTRICTS: list[DistrictRow] = {json.dumps(active_districts, ensure_ascii=False, indent=4)}
'''

    OUT.parent.mkdir(parents=True, exist_ok=True)
    OUT.write_text(body, encoding="utf-8")
    print(f"counties={len(active_counties)} districts={len(active_districts)} -> {OUT.relative_to(ROOT)}")


if __name__ == "__main__":
    main()
